warden/app/oracle.go

```diff
@@ -0,0 +1,231 @@
+		oracleMetrics,
+	)
+	app.SetPrepareProposal(proposalHandler.PrepareProposalHandler())
+	app.SetProcessProposal(proposalHandler.ProcessProposalHandler())
+
+	// Create the aggregation function that will be used to aggregate oracle data
+	// from each validator.
+	aggregatorFn := voteweighted.MedianFromContext(
+		app.Logger(),
+		app.StakingKeeper,
+		voteweighted.DefaultPowerThreshold,
+	)
+
+	// Create the pre-finalize block hook that will be used to apply oracle data
+	// to the state before any transactions are executed (in finalize block).
+	oraclePreBlockHandler := oraclepreblock.NewOraclePreBlockHandler(
+		app.Logger(),
+		aggregatorFn,
+		app.OracleKeeper,
+		oracleMetrics,
+		currencypair.NewDeltaCurrencyPairStrategy(app.OracleKeeper),
+		compression.NewCompressionVoteExtensionCodec(
+			compression.NewDefaultVoteExtensionCodec(),
+			compression.NewZLibCompressor(),
+		),
+		compression.NewCompressionExtendedCommitCodec(
+			compression.NewDefaultExtendedCommitCodec(),
+			compression.NewZStdCompressor(),
+		),
+	)
+
+	app.SetPreBlocker(oraclePreBlockHandler.PreBlocker())
+
+	// Create the vote extensions handler that will be used to extend and verify
+	// vote extensions (i.e. oracle data).
+	cps := currencypair.NewDeltaCurrencyPairStrategy(app.OracleKeeper)
+	veCodec := compression.NewCompressionVoteExtensionCodec(
+		compression.NewDefaultVoteExtensionCodec(),
+		compression.NewZLibCompressor(),
+	)
+	extCommitCodec := compression.NewCompressionExtendedCommitCodec(
+		compression.NewDefaultExtendedCommitCodec(),
+		compression.NewZStdCompressor(),
+	)
+	voteExtensionsHandler := ve.NewVoteExtensionHandler(
+		app.Logger(),
+		app.oracleClient,
+		time.Second,
+		cps,
+		veCodec,
+		aggregator.NewOraclePriceApplier(
+			aggregator.NewDefaultVoteAggregator(
+				app.Logger(),
+				aggregatorFn,
+				// we need a separate price strategy here, so that we can optimistically apply the latest prices
+				// and extend our vote based on these prices
+				currencypair.NewDeltaCurrencyPairStrategy(app.OracleKeeper),
+			),
+			app.OracleKeeper,
+			veCodec,
+			extCommitCodec,
+			app.Logger(),
+		),
+		oracleMetrics,
+	)
+	app.SetExtendVoteHandler(voteExtensionsHandler.ExtendVoteHandler())
+	app.SetVerifyVoteExtensionHandler(voteExtensionsHandler.VerifyVoteExtensionHandler())
+}
+
+type AppUpgrade struct {
+	Name         string
+	Handler      upgradetypes.UpgradeHandler
+	StoreUpgrade storetypes.StoreUpgrades
+}
+
+// createSlinkyUpgrader returns the upgrade name and an upgrade handler that:
+// - runs migrations
+// - updates the consensus keeper params with a vote extension enable height. (height of upgrade + 10).
+// - adds the core markets to x/marketmap keeper.
+// additionally, it returns the required StoreUpgrades needed for the new slinky modules added to this chain.
+func createSlinkyUpgrader(app *App) AppUpgrade {
+	return AppUpgrade{
+		Name: "TODO: NAME THIS UPGRADE",
```

Comment:
you'll want to replace this with whatever you want the slinky upgrade name to be. or i can do that in this PR. just let me know the preference!

Reply:
let's call it "v03-to-v04" please

Reply:
added name here [name the upgrade](https://github.com/warden-protocol/wardenprotocol/pull/425/commits/dc2f2a81937038cdbca8e7de0b4f36aa0a65c9f6)